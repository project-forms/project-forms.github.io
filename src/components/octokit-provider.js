// @ts-check

import {
  createContext,
  createElement,
  useContext,
  useState,
  useMemo,
  useCallback,
  useEffect,
} from "react";
import { Octokit } from "@octokit-next/core";

export const OctokitContext = createContext(
  /** @type {import('./octokit-provider').OctokitContextValue | null} */
  (null)
);

// @ts-expect-error - look into how to define env vars in Vite
const backendBaseUrl = import.meta.env.VITE_BACKEND_BASE_URL || "";

let storeData = null;
/** @type {import('./octokit-provider').Store} */
let DEFAULT_STORE = {
  get() {
    return storeData;
  },
  set(data) {
    storeData = data;
  },
};

/**
 * @param {import("./octokit-provider").Store} store
 */
async function loadData(store) {
  return store.get();
}

/**
 * @param {Octokit} octokit
 * @param {React.Dispatch<React.SetStateAction<import('./octokit-provider').AuthState>>} setAuthState
 * @param {import("./octokit-provider").Store} store
 */
async function login(octokit, setAuthState, store) {
  const { data: user } = await octokit.request("GET /user");
  setAuthState({ type: "authenticated", user, octokit });
  const { token } = await octokit.auth();
  await store.set({ user, token });
}

/**
 * @param {Octokit} octokit
 * @param {React.Dispatch<React.SetStateAction<import('./octokit-provider').AuthState>>} setAuthState
 * @param {import("./octokit-provider").Store} store
 */
async function logout(octokit, setAuthState, store) {
  await store.set(null);
  setAuthState({ type: "unauthenticated", octokit });
}

/** @type {import('./octokit-provider').AuthState} */
let INITIAL_AUTH_STATE = {
  type: "loading",
  octokit: new Octokit(),
};

/**
 * @param {React.PropsWithChildren & {store: import("./octokit-provider").Store}} props
 * @returns {ReturnType<React.createElement>}
 */
export const OctokitProvider = ({ children, store = DEFAULT_STORE }) => {
  const [authState, setAuthState] = useState(INITIAL_AUTH_STATE);

  const exportedLogout = useCallback(async () => {
    logout(authState.octokit, setAuthState, store);
  }, [authState, setAuthState, store]);

  // load initial auth state from store
  useEffect(() => {
    console.log("Check for `code` query param");
    const code = new URL(location.href).searchParams.get("code");
    if (code) {
      handleCodeInUrl(store, setAuthState, code);
      return;
    }

    console.log("load initial auth state from store");
    loadData(store).then(async (data) => {
      if (!data) {
        setAuthState({ type: "unauthenticated", octokit: new Octokit() });
        return;
      }

      const octokit = new Octokit({
        auth: data.token,
      });

      setAuthState({ type: "authenticated", user: data.user, octokit });

      try {
        await login(octokit, setAuthState, store);
      } catch {
        await logout(octokit, setAuthState, store);
      }
    });
  }, [setAuthState, store]);

  return createElement(
    OctokitContext.Provider,
    {
      value: {
        authState,
        logout: exportedLogout,
      },
    },
    children
  );
};

/**
 * @param {import('./octokit-provider').Store} store
 * @param {React.Dispatch<React.SetStateAction<import('./octokit-provider').AuthState>>} setAuthState
 * @param {string} code
 */
async function handleCodeInUrl(store, setAuthState, code) {
  // remove ?code=... from URL
  const path =
    location.pathname +
    location.search.replace(/\b(code|state)=\w+/g, "").replace(/[?&]+$/, "");
  history.replaceState({}, "", path);

  setAuthState({ type: "loading", octokit: new Octokit() });
  const token = await getTokenFromCode(code);
  const octokit = new Octokit({ auth: token });

  try {
    await login(octokit, setAuthState, store);
  } catch {
    await logout(octokit, setAuthState, store);
  }
}

/**
 * @param {string} code
 * @returns Promise<string | null>
 */
async function getTokenFromCode(code) {
  try {
    const response = await fetch(`${backendBaseUrl}/api/github/oauth/token`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ code }),
    });
    const { authentication } = await response.json();
    // TODO - store authentication.refreshToken
    return authentication.token;
  } catch (e) {
    return null;
  }
}
