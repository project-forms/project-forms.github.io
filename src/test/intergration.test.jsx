import {
  describe,
  expect,
  test,
  beforeAll,
  afterAll,
  afterEach,
  vi,
} from "vitest";
import { render, waitFor } from "@testing-library/react";
import { BaseStyles, ThemeProvider } from "@primer/react";
import { setupServer } from "msw/node";
import { http, HttpResponse, graphql } from "msw";
import { OctokitProvider } from "../components/octokit-provider.js";
import Root from "../root.jsx";
import UserResponse from "./fixtures/api.github.com/user.json";

const server = setupServer();
const githubGraphQL = graphql.link("https://api.github.com/graphql");
class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
}
describe("App", () => {
  beforeAll(() => {
    // It seems like Primer components rely on the existence of `window.ResizeObserver` and it's not (properly) mocked by JS Dom
    // TODO try this instead: vi.stubGlobal("ResizeObserver", ResizeObserver)
    window.ResizeObserver = ResizeObserver;

    // mock local store
    vi.mock("../lib/create-store.js", () => {
      let storeData = {};

      return {
        default: () => ({
          get() {
            return storeData;
          },
          set(data) {
            storeData = data;
          },
        }),
      };
    });

    server.listen();
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  test("/project-forms/demo/projects/1/issues/new?code=123", async () => {
    const mockLocation = {
      pathname: "/project-forms/demo/projects/1/issues/new",
      origin: "http://localhost.test",
      search: "?code=123",
      searchParams: {
        get: (key) => {
          if (key === "code") {
            return "123";
          }
        },
      },
      href: "http://localhost.test/project-forms/demo/projects/1/issues/new?code=123",
    };

    server.use(
      http.get("https://api.github.com/user", () => {
        return HttpResponse.json(UserResponse);
      }),

      http.post(
        "http://localhost.test/api/github/oauth/token",
        async ({ request }) => {
          const requestBody = await request.json();
          expect(requestBody).toStrictEqual({
            code: "123",
          });

          return HttpResponse.json({
            authentication: {
              token: "secret_123",
            },
          });
        }
      ),
      githubGraphQL.query("verifyAccess", ({ variables }) => {
        expect(variables).toStrictEqual({
          owner: "project-forms",
          repo: "demo",
          projectNumber: 1,
        });

        return HttpResponse.json({
          data: {
            repository: "randomRepo",
            owner: {
              project: {
                viewerCanUpdate: true,
              },
            },
          },
        });
      }),
      githubGraphQL.query("getProjectWithItems", ({ variables }) => {
        expect(variables).toStrictEqual({
          owner: "project-forms",
          number: 1,
        });

        return HttpResponse.json({
          data: {
            userOrOrganization: {
              projectV2: {
                title: "My Project",
                url: "Project URL",
                fields: {
                  nodes: [
                    {
                      name: "Field 1",
                      dataType: "text",
                      id: "1",
                    },
                    {
                      name: "Field 2",
                      dataType: "text",
                      id: "2",
                    },
                  ],
                },
              },
            },
          },
        });
      })
    );

    const { asFragment, getByText } = render(<Root />);

    await waitFor(
      () => {
        const title = getByText(/My Project/);
        expect(title).not.toBeNull();
      },
      { timeout: 1000 }
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
