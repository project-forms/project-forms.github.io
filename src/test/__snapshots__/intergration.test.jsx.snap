// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > /project-forms/demo/projects/1/issues/new?code=123 1`] = `
<DocumentFragment>
  <div
    class="BaseStyles__Base-sc-nfjs56-0 kBCNKa"
    color="fg.default"
    data-color-mode="light"
    data-dark-theme="dark"
    data-light-theme="light"
    data-portal-root="true"
    font-family="normal"
  >
    <header
      class="Header-sc-1wu4agm-0 cwcmrC"
    >
      <div
        class="Header__HeaderItem-sc-1wu4agm-1 hFzBrP"
      >
        <a
          class="Header__HeaderLink-sc-1wu4agm-2 iizeLw"
          href="https://github.com/project-forms"
        >
          <svg
            aria-hidden="true"
            class="Octicon-sc-9kayk9-0 eoNnjt"
            fill="currentColor"
            focusable="false"
            height="32"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 24 24"
            width="32"
          >
            <path
              d="M2 3.75C2 2.784 2.784 2 3.75 2h16.5c.966 0 1.75.784 1.75 1.75v16.5A1.75 1.75 0 0 1 20.25 22H3.75A1.75 1.75 0 0 1 2 20.25ZM9 9v11.5h11.25a.25.25 0 0 0 .25-.25V9Zm11.5-1.5V3.75a.25.25 0 0 0-.25-.25H9v4ZM3.5 9v11.25c0 .138.112.25.25.25H7.5V9Zm4-1.5v-4H3.75a.25.25 0 0 0-.25.25V7.5Z"
            />
          </svg>
          <span>
            Project Forms
          </span>
        </a>
      </div>
      <div
        class="Header__HeaderItem-sc-1wu4agm-1 jJZXfG"
      >
        <button
          aria-haspopup="true"
          class="types__StyledButton-sc-ws60qy-0 SQOoL"
          data-component="IconButton"
          data-no-visuals="true"
          id="react-aria-:r0:"
          tabindex="0"
          type="button"
        >
          <img
            alt=""
            class="Avatar__StyledAvatar-sc-2lv0r8-0 gNYUCl"
            height="32"
            size="32"
            src="https://github.com/images/error/octocat_happy.gif"
            width="32"
          />
        </button>
      </div>
    </header>
    <div
      class="Box-sc-g0xbh4-0 geFQUw"
    >
      <div
        class="Box-sc-g0xbh4-0 chbkDf"
      >
        <nav
          aria-label="Breadcrumbs"
          class="Breadcrumbs__BreadcrumbsBase-sc-9m4wsf-1 jHqoNi"
        >
          <ol
            class="Box-sc-g0xbh4-0 eItYAW"
          >
            <li
              class="Breadcrumbs__Wrapper-sc-9m4wsf-0 jUaWOZ"
            >
              <a
                class="Breadcrumbs__BreadcrumbsItem-sc-9m4wsf-2 cGAxOd"
                href="https://github.com/project-forms"
              >
                project-forms
              </a>
            </li>
            <li
              class="Breadcrumbs__Wrapper-sc-9m4wsf-0 jUaWOZ"
            >
              <a
                aria-current="page"
                class="Breadcrumbs__BreadcrumbsItem-sc-9m4wsf-2 cGAxOd selected"
                href="https://github.com/project-forms/demo"
              >
                demo
              </a>
            </li>
          </ol>
        </nav>
        <h2
          class="Heading__StyledHeading-sc-1c1dgg0-0 dJRcyg"
        >
          Submit to project 
          <a
            class="Link__StyledLink-sc-14289xe-0 iCnTAO"
            href="Project URL"
          >
            #1 My Project
          </a>
        </h2>
      </div>
    </div>
    <div
      class="Box-sc-g0xbh4-0 chbkDf"
    >
      <form>
        <div
          class="Box-sc-g0xbh4-0 bjtFqb"
        >
          <div
            class="Box-sc-g0xbh4-0 dgmmuv"
            display="flex"
          >
            <label
              class="_VisuallyHidden__VisuallyHidden-sc-11jhm7a-0 hjcOCW"
              for="react-aria-:r2:"
            >
              <span
                class="Box-sc-g0xbh4-0 jZSkdQ"
                display="flex"
              >
                <div
                  class="Box-sc-g0xbh4-0 bhnSKu"
                >
                  Issue title
                </div>
                <span>
                  *
                </span>
              </span>
            </label>
            <span
              aria-busy="false"
              class="TextInputWrapper__TextInputBaseWrapper-sc-1mqhpbi-0 TextInputWrapper-sc-1mqhpbi-1 bvOKsi jfWWKv TextInput-wrapper"
            >
              <input
                aria-describedby=""
                class="UnstyledTextInput-sc-14ypya-0 eHKHqO"
                data-component="input"
                id="react-aria-:r2:"
                name="title"
                required=""
                type="text"
              />
            </span>
          </div>
          <div
            class="Box-sc-g0xbh4-0 dgmmuv"
            display="flex"
          >
            <label
              class="_VisuallyHidden__VisuallyHidden-sc-11jhm7a-0 hjcOCW"
              for="react-aria-:r3:"
            >
              Issue description
            </label>
            <span
              class="TextInputWrapper__TextInputBaseWrapper-sc-1mqhpbi-0 bvOKsi"
            >
              <textarea
                aria-describedby=""
                aria-invalid="false"
                aria-required="false"
                class="Textarea__StyledTextarea-sc-1lf8it-0 jqAtDs"
                cols="30"
                id="react-aria-:r3:"
                name="body"
                rows="7"
              />
            </span>
          </div>
          <div
            class="Box-sc-g0xbh4-0 fPbvzj"
            display="flex"
          >
            <button
              class="types__StyledButton-sc-ws60qy-0 cRArGu"
              data-no-visuals="true"
              disabled=""
              type="submit"
            >
              <span
                class="Box-sc-g0xbh4-0 gUkoLg"
                data-component="buttonContent"
              >
                <span
                  data-component="text"
                >
                  Submit new issue
                </span>
              </span>
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
